TestFoundation/TestDateFormatter.swift

```diff
@@ -356,4 +358,54 @@ class TestDateFormatter: XCTestCase {
         // Time zone should go back to the system one.
         XCTAssertEqual(f.timeZone, NSTimeZone.system)
     }
+
+    func test_setTimeZone() {
+        // Test two different time zones. Should ensure that if one
+        // happens to be TimeZone.current, we still get a valid test.
+        let newYork = TimeZone(identifier: "America/New_York")!
+        let losAngeles = TimeZone(identifier: "America/Los_Angeles")!
+
+        XCTAssertNotEqual(newYork, losAngeles)
+
+        // Case 1: New York
+        let f = DateFormatter()
+        f.timeZone = newYork
+        XCTAssertEqual(f.timeZone, newYork)
+
+        // Case 2: Los Angeles
+        f.timeZone = losAngeles
+        XCTAssertEqual(f.timeZone, losAngeles)
+    }
+
+    func test_ExpectedTimeZone() {
+        let gmt = TimeZone(abbreviation: DEFAULT_TIMEZONE)
+        let newYork = TimeZone(identifier: "America/New_York")!
+        let losAngeles = TimeZone(identifier: "America/Los_Angeles")!
+
+        XCTAssertNotEqual(newYork, losAngeles)
+
+        let now = Date()
+
+        let f = DateFormatter()
+        f.dateFormat = "z"
+
+        // Case 1: TimeZone.current
+        f.timeZone = TimeZone.current
+        XCTAssertEqual(f.string(from: now), TimeZone.current.abbreviation())
+
+        // Case 2: New York
+        f.timeZone = newYork
+        XCTAssertEqual(f.string(from: now), newYork.abbreviation())
+
+        // Case 3: Los Angeles
+        f.timeZone = losAngeles
+        XCTAssertEqual(f.string(from: now), losAngeles.abbreviation())
+
+        guard gmt != TimeZone.current else {
```

Comment:
Should there be another test case after this guard for GMT? (If not, I don't see much of a reason to keep the print statements around.)

Reply:
I really want an XCTInconclusive(). The issue is that because TimeZone.current gets treated as GMT when it isn’t, if you happen to be using GMT, this test can’t tell you conclusively that the bug doesn’t exist.

Reply:
So I guess my question is, should this test just pass in this case without the “warnings”? Should a more specific test be found checking TimeZone.current’s identifier string so this test doesn’t need to test for the fallout of that case?

It’s kinda why I’m not a huge fan of how I wrote this, and was hoping for pointers on what would be more acceptable.

Reply:
Assume TimeZone.current is "America/Los_Angeles"

You can ask it's identifier, and it's abbreviation and it will give you what you expect. However, because of this:

```
CFStringCreateWithBytes(kCFAllocatorSystemDefault, (uint8_t *)linkbuf + zoneInfoDirLen,
                                            strlen(linkbuf) - zoneInfoDirLen + 1, kCFStringEncodingUTF8, false);
```

Which allocates one byte too many, and includes the null terminator in the CFString, if I then give DateFormatter this malformed TimeZone, I get GMT, despite it being PDT/PST. It also affects the intepretation of strings in both directions. Instead of getting say, "Midnight, PST", I get "7-8am, GMT". 

i.e. If you have a malformed TimeZone, DateFormatter assumes that it is GMT, not your actual time zone. So if you are trying to format strings that implicitly assume local time, "08:00" for example, then you get bunk results from DateFormatter that throw your time off by your GMT offset.

Reply:
Either way, I'm going to try to write a more focused test that checks for the bad length, which means the warning can become a comment on case 1 as information for folks down the road wondering why non-GMT time zones are suddenly being treated as GMT by DateFormatter for no apparent reason.

Reply:
There's one more subtle wrinkle here:

`TimeZone(identifier: "America/Los_Angeles")`: Works
`TimeZone(identifier: TimeZone.current.identifier)`: Works
`TimeZone.current`: Broken

The extra character was only being appended in CFCreateSystemTimeZone(), and so it doesn't affect you if you can declare the time zone directly.

Does ICU take in the string length? That would explain it if it relies on the buffer length and not a terminator like CFString.

Reply:
ICU _can_ search for a NUL-terminating character if you tell it to, but `CFStringRef` does not require NUL termination; we pass the contents of the `CFStringRef` along with its length directly to ICU, so if the `CFStringRef` contains a NUL character, ICU will get that. It appears that ICU reads this NUL character in a way that makes `"America/Los_Angeles\0" != "America/Los_Angeles"`.

Reproduces easily with

```objc
#import <Foundation/Foundation.h>
#include <unicode/udat.h>

int main()
{
    @autoreleasepool {
#define U_CHECK(status, message) if (U_FAILURE(status)) { NSLog(message @": %s", u_errorName(status)); exit(EXIT_FAILURE); }
        UErrorCode status = U_ZERO_ERROR;
        UDateFormat *formatter = udat_open(UDAT_PATTERN, UDAT_PATTERN, "en_US_POSIX", u"America/Los_Angeles\0", 20, u"z", -1, &status);
        U_CHECK(status, @"Unable to create date formatter");
        
        UChar result[BUFSIZ] = {0};
        int32_t length = udat_format(formatter, ucal_getNow(), result, BUFSIZ, NULL, &status);
        U_CHECK(status, @"Unable to format");
        
        NSLog(@"%@", [[NSString alloc] initWithCharactersNoCopy:result length:length freeWhenDone:NO]);
    }
}
```

producing `GMT`. Changing the length from `20` to `19` produces `PDT` as expected.

That being said: are you saying that `TimeZone.current` is _still_ broken after this change, or was broken only before the change? I'm trying to figure out what the state of things are with this fix.

Reply:
With this PR, it is no longer broken. 

The point of this test was to try to prevent a similar regression that breaks DateFormatter in a similar way.

Reply:
This seems reasonable for now — I think in the future we can just add a test to see if `TimeZone.current.identifier != "GMT` but `formatter.string(from: date) == "GMT"` to catch this too, but we shouldn't need to hold up this PR any further over that.